pkg/datasource/opensergo/opensergo.go

```diff
@@ -0,0 +1,135 @@
+
+	return ds, nil
+}
+
+func (ds *OpenSergoDataSource) Close() {
+	subscribersAll := ds.client.GetSubscriberRegistry().GetSubscribersAll()
+	subscribersAll.Range(func(key, value interface{}) bool {
+		ds.client.UnsubscribeConfig(key.(subscribe.SubscribeKey))
+		logging.Info(fmt.Sprintf("Unsubscribing OpenSergo config for target: %v", key))
+		return true
+	})
+}
+
+// Initialize
+//
+// 1. Set the handler for sentinel, to update sentinel local cache when the data from opensego was changed.
+//
+// 2. Start the NewOpenSergoClient.
+//
+// 3. Resister OpenSergo Subscribers by params
+func (ds *OpenSergoDataSource) Initialize() error {
+	ds.opensergoRuleAggregator.setSentinelUpdateHandler(ds.doUpdate)
+	ds.client.Start()
+
+	// TODO to add datasource-params in NewOpenSergoDataSource to decide register which subscribers for datasource
+	// TODO add the deciding logic in follow
+	ds.RegisterSubscribeInfoOfFaulttoleranceRule()
+	ds.RegisterSubscribeInfoOfFlowRuleStrategy()
+	return nil
+}
+
+func (ds OpenSergoDataSource) doUpdate() {
+	bytes, err := ds.ReadSource()
+	if err != nil {
+		logging.Warn("[OpenSergo] Succeed to read source in Initialize()", "namespace", ds.namespace, "app", ds.app, "content", fmt.Sprintf(string(bytes)))
+	}
+
+	ds.Handle(bytes)
+}
+
+func (ds *OpenSergoDataSource) ReadSource() ([]byte, error) {
+	// assemble updated MixedRule
+	mixedRule := new(MixedRule)
+	if ds.opensergoRuleAggregator.mixedRuleCache.updateFlagMap[RuleType_FlowRule] {
+		mixedRule.FlowRule = ds.opensergoRuleAggregator.mixedRuleCache.FlowRule
+	}
+	// TODO assembler other rule-type
+
+	logging.Info("[OpenSergo] Succeed to read source", "namespace", ds.namespace, "app", ds.app, "content", mixedRule)
+	bytes, err := json.Marshal(mixedRule)
+	if err != nil {
+		return nil, err
+	}
+	return bytes, nil
+}
+
+func (ds *OpenSergoDataSource) RegisterSubscribeInfoOfFaulttoleranceRule() {
+	// registry SubscribeInfo of FaultToleranceRule
+	faultToleranceRuleSubscribeKey := subscribe.NewSubscribeKey(ds.namespace, ds.app, configkind.ConfigKindRefFaultToleranceRule{})
+	faultToleranceRuleSubscribeInfo := client.NewSubscribeInfo(*faultToleranceRuleSubscribeKey)
+	faulttoleranceRuleSubscriber := NewFaulttoleranceRuleSubscriber(ds.opensergoRuleAggregator)
+	faultToleranceRuleSubscribeInfo.AppendSubscriber(faulttoleranceRuleSubscriber)
+	// log data for test
+	faultToleranceRuleSubscribeInfo.AppendSubscriber(subscribe.DefaultSubscriber{})
+	ds.client.RegisterSubscribeInfo(faultToleranceRuleSubscribeInfo)
+	logging.Info(fmt.Sprintf("Subscribing OpenSergo base fault-tolerance rules for target <%v, %v>", ds.namespace, ds.app))
+}
+
+func (ds *OpenSergoDataSource) RegisterSubscribeInfoOfFlowRuleStrategy() {
+	// registry SubscribeInfo of RateLimitStrategy
+	rateLimitStrategySubscribeKey := subscribe.NewSubscribeKey(ds.namespace, ds.app, configkind.ConfigKindRefRateLimitStrategy{})
+	rateLimitStrategySubscribeInfo := client.NewSubscribeInfo(*rateLimitStrategySubscribeKey)
```

Comment:
Same as above

Reply:
has not changed. Because of modification is not involving `rateLimitStrategySubscribeKey `.

Reply:
optimized.